tests/test-update-install-flatpaks.c

```diff
@@ -3906,6 +3906,338 @@ test_update_uninstall_flatpaks_on_reboot (EosUpdaterFixture *fixture,
+                                                                 NULL,
+                                                                 "test-repo",
+                                                                 TRUE,
+                                                                 "app/org.test.Test/arch/stable",
+                                                                 NULL));
+
+  /* Set up repos */
+  g_hash_table_insert (flatpak_repo_infos,
+                       g_strdup ("test-repo"),
+                       flatpak_repo_info_new ("test-repo",
+                                              "com.endlessm.TestInstallFlatpaksCollection",
+                                              "com.endlessm.TestInstallFlatpaksCollection"));
+
+  eos_test_setup_flatpak_repo (updater_directory,
+                               flatpak_install_infos,
+                               flatpak_repo_infos,
+                               gpg_key_file,
+                               keyid,
+                               &error);
+
+  /* Update the server, so it has a new commit (1).
+   */
+  etc_update_server (data, 1);
+  /* Update the client, so it also has a new commit (1); and, at this
+   * point, two deployments - old one pointing to commit 0 and a new
+   * one pointing to commit 1.
+   */
+  etc_update_client (data);
+
+  /* Now simulate a reboot by running eos-updater-flatpak-installer */
+  deployment_csum = get_checksum_for_deploy_repo_dir (deployment_repo_dir,
+                                                      refspec,
+                                                      &error);
+  g_assert_no_error (error);
+
+  eos_test_run_flatpak_installer (data->client->root,
+                                  deployment_csum,
+                                  default_remote_name,
+                                  &error);
+  g_assert_no_error (error);
+
+  /* Update the server again with commit 2, then update the client. This will
+   * pull in a new ref-action list with the uninstall action. */
+  etc_update_server (data, 2);
+  etc_update_client (data);
+
+  /* Reboot again, run the flatapk installer over this new commit */
+  deployment_csum = get_checksum_for_deploy_repo_dir (deployment_repo_dir,
+                                                      refspec,
+                                                      &error);
+  g_assert_no_error (error);
+
+  eos_test_run_flatpak_installer (data->client->root,
+                                  deployment_csum,
+                                  default_remote_name,
+                                  &error);
+  g_assert_no_error (error);
+
+  /* Extension should not be installed here */
+  deployed_flatpaks = eos_test_get_installed_flatpaks (updater_directory, &error);
+  g_assert_no_error (error);
+
+  g_assert_false (g_strv_contains ((const gchar * const *) deployed_flatpaks,
+                                   "org.test.Test.Extension"));
```

Comment:
You should probably also verify in the middle step that the app and its extension are installed.

Reply:
You should also verify here that the app is uninstalled.

Reply:
Hmm, my preference is that tests have a single assertion set at the end, but I think in this case I'm willing to break that rule (we should make sure that the app and extension got preinstalled by the framework)

Reply:
OOI, why do you hold that preference? That would imply that if you want to test a multi-stage process, you’d need several tests, each one going one stage further than the previous and then doing its assertions.

Reply:
See also: http://xunitpatterns.com/Assertion%20Roulette.html

Reply:
That testing pattern has some merit, but I’m not convinced it’s an improvement here.

Firstly, when `g_assert()` fails, you get the line number of the failure, so locating the failed assertion is generally trivial (unless it’s in a loop — then you should probably use `g_test_message()` at the start of each loop iteration to print the iterator).

Secondly, in this situation you’re trying to test that at the end both the app and its extension are uninstalled. If you take the pattern above literally, that means doing the test once to assert that the app is uninstalled, and doing it again to assert that the extension is uninstalled, which will double the amount of code and double the runtime.

Thirdly, adding assertions part-way through is essentially one half of asserting that the installation state of the app has changed. It’s protecting against the failure mode where (somehow) the app doesn’t get installed, and hence the final assertion that the app is uninstalled is trivially true (but for the wrong reasons). It’s kind of a take on “who tests the tests?”.

That pattern does have merit in moderation, though, and I’ve seen many tests which essentially test the entire functionality of an object in a single (very long) test case. That’s something to avoid like the plague; not because it’s hard to work out which assertion failed, but rather because it’s hard to work out what state the object under test was in at that point in the test, and whether all states of the object have actually been covered by that monster test case.

Reply:
(Sorry for the essay.)

Reply:
Oh weird, my last comment here must not have been saved here. Anyway, too many assertions in a test indicates that a test is testing too much or that it probably could use a nicer fixture (which has its own internal assertions validating that the fixture works correctly). See http://xunitpatterns.com/Assertion%20Roulette.html

Reply:
Anyway, I added assertions for now.

Reply:
I basically agree with you, though I had the following thoughts:

> Secondly, in this situation you’re trying to test that at the end both the app and its extension are uninstalled. If you take the pattern above literally, that means doing the test once to assert that the app is uninstalled, and doing it again to assert that the extension is uninstalled, which will double the amount of code and double the runtime.

If one were to take xunit-test-patterns to the extreme the correct solution to this would be a matcher that checks that a subset of flatpaks are in the list. Unfortunately matchers are highly polymorphic and that doesn't mix with C :( (I tried doing this in the Gjs unit tests, it wasn't pretty).

> Thirdly, adding assertions part-way through is essentially one half of asserting that the installation state of the app has changed. It’s protecting against the failure mode where (somehow) the app doesn’t get installed, and hence the final assertion that the app is uninstalled is trivially true (but for the wrong reasons). It’s kind of a take on “who tests the tests?”.

Yeah, usually you'd have assertions in the fixture itself. Google Test has a distinction between fatal and non-fatal assertions: https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#basic-assertions - in a past life I used the fatal assertions to test the test fixtures and non-fatal assertions to test the actual code.